concurrency/MpmcQueue.h

```diff
@@ -5,112 +5,164 @@
+          _elements(reinterpret_cast<Entry*>(memory::page::allocate(_blockSize)))
     {
-        _readCursor = 0;
-        _writeCursor = 0;
-        assert(0x100000000ull % _maxElements == 0); // "size % 2^32 must be zero";
+        assert(_capacity > 7);
+        assert(capacity < 0x80000000u);
+        assert(_capacity % _entriesPerCacheline == 0);
+
+        _readCursor = VersionedIndex();
+        _writeCursor = VersionedIndex();
 
         _cacheLineSeparator1[0] = 0;
         _cacheLineSeparator2[0] = 0;
 
-        _elements = reinterpret_cast<Entry*>(memory::page::allocate(_blockSize));
-
-        for (uint32_t i = 0; i < _maxElements; ++i)
+        for (uint32_t i = 0; i < _capacity; ++i)
         {
             new (&_elements[i]) Entry();
         }
     }
 
     ~MpmcQueue()
     {
-        for (uint32_t i = 0; i < _maxElements; ++i)
+        for (uint32_t i = 0; i < _capacity; ++i)
         {
+            if (_elements[i].state.load().state == CellState::State::committed)
+            {
+                _elements[i].value().~T();
+            }
             _elements[i].~Entry();
         }
 
         memory::page::free(_elements, _blockSize);
     }
 
-    // return false if empty
     bool pop(T& target)
     {
-        uint32_t pos = _readCursor;
-        for (;;)
+        for (auto pos = _readCursor.load(std::memory_order_consume);;)
         {
             if (!isReadable(pos))
             {
-                if (pos == _readCursor)
+                const auto readCursor = _readCursor.load(std::memory_order_consume);
+                if (pos == readCursor)
                 {
                     return false;
                 }
-                pos = _readCursor;
+                pos = readCursor;
                 continue;
             }
 
-            if (_readCursor.compare_exchange_weak(pos, pos + 1))
+            const auto newPos = nextPosition(pos);
+
+            if (_readCursor.compare_exchange_weak(pos, newPos, std::memory_order_seq_cst))
             {
                 // we own the read position now. If thread pauses here,
                 // A writer cannot write due to committed state and will not increase write cursor.
                 // A reader fail to update readCursor and has to try next slot.
-                // A reader that wrapped will not pass readable test because the pos == writepos
-                auto& entry = _elements[pos % _maxElements];
-                target = std::move(entry.value);
-                entry.state.store(CellState::emptySlot, std::memory_order_release);
+                // A reader that wrapped will not pass readable test because the readcursor version does not match cell
+                // state version
+                auto& entry = _elements[indexTransform(pos)];
+                target = std::move(entry.value());
+                entry.value().~T();
+                entry.state.store(CellState{pos.version + 1, CellState::emptySlot}, std::memory_order_release);
```

Comment:
This is the Achilles heel of MpmcQueue. There is a lock-read-release phase here that can stop writers from continuing.
If the element would have been removed from the queue in one atomic operation, the reader would either succeed in acquiring the object or fail to retry. Instead of the item containing value, the queue should have contained a versioned pointer.